pyDrop/clustering/course_graining.py

```diff
@@ -0,0 +1,249 @@
+        if n_defined_bins == 0:
+            print(f"no axes defined, using default ModuloBins bins for coarse-graining")
+            self.fit_uniform_coarse_grain(n_features)
+        elif n_defined_bins < n_features:
+            print(f"ambiguous bins definition: {n_defined_bins} axes defined for bins but\
+                  there are {n_features} total axes.")
+        
+        X_grained_bins = self.bins.value_to_id(X)
+        X_grained_bins = np.unique(X_grained_bins, axis=0)
+        X_grained_centers = self.bins.id_to_bin_center(X_grained_bins)
+
+        return X_grained_centers
+    
+    def fit(self, X):
+        X_grained_centers = self.coarse_grain(X)
+        self.coarse_model.fit(X_grained_centers)
+
+    def predict(self, X, model="coarse"):
+        y_pred = None
+        if model=="coarse":
+            y_pred = self.coarse_model.predict(X)
+        elif model=="default":
+            y_pred = self.model.fit_predict(X)
+        else:
+            raise Exception("model must be coarse or default")
+        return y_pred
+    
+    def scores(self, X, y_true, model="fine"):
+        y_pred = self.predict(X, model=model)
+        r_score = rand_score(y_true, y_pred)
+        scores = {"rand_score": r_score,
+                  "homogeneity_score": homogeneity_score(y_true, y_pred),
+                  "completeness_score": completeness_score(y_true, y_pred),
+                  "n_misclassified": len(y_true)*(1-r_score)
+                 }
+        return scores
+
+class KNNCalico(CGCluster):
+    def __init__(self, bins=Bins(), k_means_model=KMeans()):
+        """
+
+        """
+        # initialize data and the expected number of clusters 
+        self.bins = bins
+        self.model = k_means_model
+
+        self.coarse_model = deepcopy(k_means_model)
+        self.coarse_model.set_params(n_init='auto')
+
+        self.fine_model = deepcopy(k_means_model)
+        self.fine_model.set_params(n_init=1)
+
+    def fit(self, X):
+        X_grained_centers = self.coarse_grain(X)
+        self.coarse_model.fit(X_grained_centers)
+        coarse_centers = self.coarse_model.cluster_centers_
+
+        self.fine_model.set_params(init=coarse_centers)
+        self.fine_model.fit(X)
+
+    def predict(self, X, model="fine", return_centers=False):
+        y_pred = None
+        centers = None
+        if model=="fine":
+            y_pred = self.fine_model.predict(X)
+            centers = self.fine_model.cluster_centers_
+        elif model=="coarse":
+            y_pred = self.coarse_model.predict(X)
+            centers = self.coarse_model.cluster_centers_
+        elif model=="default":
+            y_pred = self.model.fit_predict(X)
+            centers = self.model.cluster_centers_
+        else:
+            raise Exception("model must be fine or coarse")
+        if return_centers:
+            return y_pred, centers
+        else:
+            return y_pred
+    
+if __name__ == "__main__":
```

Comment:
Added some files to the tests/ directory for current tests and any future tests from the rest of the team. Uses pytest since that is what I am familiar with at the moment. I will look into creating a CLI and Defopt but am unfamiliar.